src/main/java/uk/gov/companieshouse/chs/gov/uk/notify/integration/api/templatepersonalisation/TemplatePersonaliser.java

```diff
@@ -0,0 +1,104 @@
+package uk.gov.companieshouse.chs.gov.uk.notify.integration.api.templatepersonalisation;
+
+import static org.apache.commons.lang.StringUtils.isEmpty;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+import org.springframework.stereotype.Component;
+import org.thymeleaf.ITemplateEngine;
+import org.thymeleaf.context.Context;
+import uk.gov.companieshouse.api.chs.notification.model.Address;
+import uk.gov.companieshouse.chs.gov.uk.notify.integration.api.exception.LetterValidationException;
+import uk.gov.companieshouse.chs.gov.uk.notify.integration.api.templatelookup.ChLetterTemplate;
+import uk.gov.companieshouse.chs.gov.uk.notify.integration.api.validation.TemplateContextValidator;
+
+@Component
+public class TemplatePersonaliser {
+
+    private final ITemplateEngine templateEngine;
+    private final TemplateContextValidator validator;
+
+    public TemplatePersonaliser(ITemplateEngine templateEngine,
+                                TemplateContextValidator validator) {
+        this.templateEngine = templateEngine;
+        this.validator = validator;
+    }
+
+    /**
+     * Populates the letter Thymeleaf template with the data for the letter.
+     * @param template the {@link ChLetterTemplate} identifying the template to be used
+     * @param  personalisationDetails the {@link Map} providing the data to be substituted into the
+     *               letter template substitution variables
+     * @return the HTML representation of the letter
+     */
+    public String personaliseLetterTemplate(ChLetterTemplate template,
+                                            Map<String, String> personalisationDetails,
+                                            Address address) {
+
+        var context = new Context();
+
+        // Use today's date for traceability.
+        var format = DateTimeFormatter.ofPattern("dd MMMM yyyy");
+        context.setVariable("date", LocalDate.now().format(format));
+
+        var upperCaseCompanyName = getUpperCasedCompanyName(personalisationDetails);
+        populateAddress(context, address, upperCaseCompanyName);
+        personaliseLetter(context, personalisationDetails, upperCaseCompanyName);
+
+        validator.validateContextForTemplate(context, template);
+
+        return templateEngine.process(template.id(), context);
+    }
+
+    private String getUpperCasedCompanyName(Map<String, String> personalisationDetails) {
+        // Company name must be provided and is always rendered in UPPER CASE in the letter.
+        if (isEmpty(personalisationDetails.get("company_name"))) {
+            throw new LetterValidationException(
+                    "No company name found in the letter personalisation details.");
+        }
+        return personalisationDetails.get("company_name").toUpperCase();
```

Comment:
We could use `isBlank` instead to also capture whitespace strings, which basically isn't a company name? Could also maybe extract the name to a variable so you're not pulling it from the map twice, but that's minor, feel free to ignore

```suggestion
        String companyName = personalisationDetails.get("company_name");
        if (isBlank(companyName)) {
            throw new LetterValidationException(
                    "No company name found in the letter personalisation details.");
        }
        return companyName.toUpperCase();
```